File: modules/cdp-crawler/ad-heatmap.js
/**
 * Ad Heatmap Generator
 * Detects ads per scroll level and calculates MFA indicators
 */

const logger = require('../logger');

class AdHeatmapGenerator {
  constructor(client) {
    this.client = client;
    this.Runtime = client.Runtime;
    this.DOM = client.DOM;
  }

  // Common ad selectors
  static AD_SELECTORS = [
    // Google Ads / AdX / Ad Manager / GPT (comprehensive)
    '[id*="google_ads"]', '[id*="gpt-"]', '[class*="adunit"]',
    '[data-google-query-id]', 'ins.adsbygoogle', '[id*="div-gpt-ad"]',
    '[id*="google_ads_iframe"]', 'iframe[id*="google_ads"]',
    'iframe[src*="googleads"]', 'iframe[src*="tpc.googlesyndication"]',
    'iframe[src*="pagead2.googlesyndication"]', 'iframe[src*="securepubads"]',
    '[data-ad-slot]', '[data-ad-client]', '[data-ad-format]',
    '[class*="adsbygoogle"]', '[id*="aswift"]',
    'iframe[name*="google_ads"]', 'iframe[title*="Advertisement"]',
    '[class*="google-ad"]', '[id*="google-ad"]',
    'div[data-google-container-id]', '[data-text-ad]',

    // Google AdX / Ad Exchange specific
    'iframe[id*="aswift_"]', 'iframe[name*="aswift_"]',
    '[id*="google_image_div"]', 'div[id*="ad_unit"]',
    'iframe[src*="safeframe"]', '[class*="safeframe"]',
    '[id*="sf_"]', 'iframe[name*="safeframe"]',
    '[data-load-complete]', '[data-google-av-cxn]',
    'div[id*="_ads_"]', '[class*="GoogleActiveViewElement"]',
    'iframe[src*="googleadservices"]', '[id*="google_image_"]',
    '[id*="ad-div"]', '[class*="ad-wrapper"]',

    // Generic ad containers
    '[class*="ad-slot"]', '[class*="advertisement"]', '[class*="ad-container"]',
    '[class*="ad-wrapper"]', '[class*="ad-banner"]', '[data-ad]',

    // Ad iframes
    'iframe[src*="doubleclick"]', 'iframe[src*="googlesyndication"]',
    'iframe[src*="amazon-adsystem"]', 'iframe[id*="google_ads"]',

    // Native ads
    '[class*="taboola"]', '[class*="outbrain"]', '[class*="mgid"]',
    '[class*="revcontent"]', '[id*="taboola"]', '[id*="outbrain"]',

    // Video ads
    '[class*="video-ad"]', '[class*="preroll"]', '[class*="midroll"]',

    // Generic ID/Class patterns (common in MFA)
    '[id^="ad-"]', '[id^="banner-"]', '[id^="div-gpt-"]',
    '[class^="ad-"]', '[class^="banner-"]', '[class*="sponsored"]',
    '[aria-label="Advertisement"]', '[aria-label="Sponsored"]',

    // Specific Bidders/Networks
    '[id*="criteo"]', '[class*="criteo"]',
    '[id*="pubmatic"]', '[class*="pubmatic"]',
    '[id*="rubicon"]', '[class*="rubicon"]',
    '[id*="openx"]', '[class*="openx"]',
    '[id*="amazon"]', '[class*="amzn"]',

    // === NEW: Deceptive Download Ads (MFA favorite) ===
    '[class*="download"]', '[id*="download"]',
    'a[href*="download"][class*="button"]', 'button[class*="download"]',
    '[class*="fake-download"]', '[class*="dl-button"]',

    // Fake system notifications / alerts
    '[class*="notification-ad"]', '[class*="alert-ad"]',
    '[class*="system-message"]', '[class*="update-notification"]',

    // Pop-under / overlay patterns
    '[class*="overlay-ad"]', '[class*="popup-ad"]', '[class*="modal-ad"]',
    '[class*="interstitial"]', '[class*="splash-ad"]',

    // AdSense alternatives commonly used by MFA
    '[class*="propeller"]', '[id*="propeller"]',
    '[class*="adsterra"]', '[id*="adsterra"]',
    '[class*="popcash"]', '[id*="popcash"]',
    '[class*="admaven"]', '[id*="admaven"]',
    '[class*="adcash"]', '[id*="adcash"]',
    '[class*="monetag"]', '[id*="monetag"]',
    '[class*="hilltopads"]', '[id*="hilltopads"]',

    // Widget / recommendation areas (often ads)
    '[class*="widget-ad"]', '[class*="sidebar-ad"]',
    '[class*="related-ad"]', '[class*="promoted"]',

    // Data attributes often used by ad networks
    '[data-ad-slot]', '[data-ad-client]', '[data-ad-format]',
    '[data-adblockkey]', '[data-cfasync]'
  ];

  async detectAdsInViewport() {
    const { result } = await this.Runtime.evaluate({
      expression: `
        (() => {
          const ads = [];
          const viewportHeight = window.innerHeight;
          const viewportWidth = window.innerWidth;
          const scrollY = window.scrollY;
          
          const selectors = ${JSON.stringify(AdHeatmapGenerator.AD_SELECTORS)};
          
          selectors.forEach(selector => {
            try {
              document.querySelectorAll(selector).forEach(el => {
                const rect = el.getBoundingClientRect();
                const computedStyle = window.getComputedStyle(el);
                
                // Only include visible elements
                if (rect.height > 10 && rect.width > 10 && 
                    computedStyle.display !== 'none' &&
                    computedStyle.visibility !== 'hidden') {
                  
                  const inViewport = rect.bottom > 0 && rect.top < viewportHeight;
                  const isAboveFold = rect.top < viewportHeight && scrollY === 0;
                  
                  // Check if it's likely an ad iframe
                  const isIframe = el.tagName === 'IFRAME';
                  const src = el.src || '';
                  const hasAdSrc = isIframe && (
                    src.includes('doubleclick') ||
                    src.includes('googlesyndication') ||
                    src.includes('googleads') ||
                    src.includes('amazon-adsystem') ||
                    src.includes('adsystem') ||
                    src.includes('ad.') ||
                    src.includes('/ads/') ||
                    src.includes('pagead') ||
                    src.includes('adservice') ||
                    src.includes('securepubads') ||
                    src.includes('tpc.google')
                  );
                  
                  ads.push({
                    selector: selector,
                    tagName: el.tagName,
                    id: el.id || null,
                    className: el.className || null,
                    x: Math.round(rect.x),
                    y: Math.round(rect.y + scrollY),
                    width: Math.round(rect.width),
                    height: Math.round(rect.height),
                    area: Math.round(rect.width * rect.height),
                    inViewport,
                    isAboveFold,
                    isIframe,
                    hasAdSrc
                  });
                }
              });
            } catch (e) {
              // Selector failed, skip
            }
          });
          
          // === NEW: Text-based heuristic detection for deceptive elements ===
          const deceptivePatterns = [
            /download\\s*(now|free|here|button|right now)/i,
            /click\\s*(here|to|now)/i,
            /free\\s*(download|install|get)/i,
            /install\\s*(now|free)/i,
            /update\\s*(required|now|available)/i,
            /your\\s*(system|computer|device)/i,
            /virus\\s*(detected|found|alert)/i,
            /warning[:\\s]/i,
            /congratulations/i
          ];
          
          // Check buttons and prominent links for deceptive text
          document.querySelectorAll('button, a.button, a[class*="btn"], [class*="button"], [role="button"]').forEach(el => {
            const text = (el.innerText || '').trim().toLowerCase();
            const isDeceptive = deceptivePatterns.some(pattern => pattern.test(text));
            
            if (isDeceptive) {
              const rect = el.getBoundingClientRect();
              const computedStyle = window.getComputedStyle(el);
              
              if (rect.height > 10 && rect.width > 10 && 
                  computedStyle.display !== 'none' &&
                  computedStyle.visibility !== 'hidden') {
                
                ads.push({
                  selector: 'DECEPTIVE_TEXT: ' + text.substring(0, 30),
                  tagName: el.tagName,
                  id: el.id || null,
                  className: el.className || null,
                  x: Math.round(rect.x),
                  y: Math.round(rect.y + scrollY),
                  width: Math.round(rect.width),
                  height: Math.round(rect.height),
                  area: Math.round(rect.width * rect.height),
                  inViewport: rect.bottom > 0 && rect.top < viewportHeight,
                  isAboveFold: rect.top < viewportHeight && scrollY === 0,
                  isIframe: false,
                  hasAdSrc: false,
                  isDeceptive: true
                });
              }
            }
          });
          
          // Deduplicate by position (same ad matched by multiple selectors)
          const unique = [];
          const seen = new Set();
          ads.forEach(ad => {
            const key = ad.x + ',' + ad.y + ',' + ad.width + ',' + ad.height;
            if (!seen.has(key)) {
              seen.add(key);
              unique.push(ad);
            }
          });
          
          return unique;
        })()
      `,
      returnByValue: true
    });

    return result.value || [];
  }

  async measureCLS() {
    const { result } = await this.Runtime.evaluate({
      expression: `
        new Promise(resolve => {
          let cls = 0;
          try {
            const observer = new PerformanceObserver((list) => {
              for (const entry of list.getEntries()) {
                if (!entry.hadRecentInput) {
                  cls += entry.value;
                }
              }
            });
            observer.observe({ type: 'layout-shift', buffered: true });
            setTimeout(() => {
              observer.disconnect();
              resolve(cls);
            }, 1000);
          } catch (e) {
            resolve(0);
          }
        })
      `,
      awaitPromise: true,
      returnByValue: true
    });

    return result.value || 0;
  }

  async getPageDimensions() {
    const { result } = await this.Runtime.evaluate({
      expression: `({
        viewportHeight: window.innerHeight,
        viewportWidth: window.innerWidth,
        totalHeight: document.body.scrollHeight,
        scrollY: window.scrollY
      })`,
      returnByValue: true
    });
    return result.value;
  }

  async captureLevel(scrollY, viewportHeight, levelIndex) {
    const ads = await this.detectAdsInViewport();
    const cls = await this.measureCLS();

    // Calculate viewport area
    const viewportArea = viewportHeight * 1920; // Assume 1920 width

    // Calculate total ad area in viewport
    const inViewportAds = ads.filter(a => a.inViewport);
    const totalAdArea = inViewportAds.reduce((sum, ad) => sum + ad.area, 0);

    // Ad density = ratio of ad pixels to viewport pixels
    const adDensity = totalAdArea / viewportArea;

    return {
      scrollY,
      levelIndex,
      viewportHeight,
      ads: inViewportAds,
      adCount: inViewportAds.length,
      totalAdArea,
      adDensity,
      cls,
      adsAboveFold: levelIndex === 0 ? ads.filter(a => a.isAboveFold).length : 0
    };
  }

  /**
   * Generate full page heatmap with scroll
   * @param {HumanSimulator} humanSimulator - Human simulator instance
   * @param {Function} onScrollLevel - Optional callback for each scroll level (e.g., for progressive content extraction)
   */
  async generateFullHeatmap(humanSimulator, onScrollLevel = null) {
    logger.info('[AdHeatmap] Starting full page heatmap generation...');

    const levels = [];

    await humanSimulator.scrollAndCapture(async (scrollY, viewportHeight, levelIndex) => {
      const levelData = await this.captureLevel(scrollY, viewportHeight, levelIndex);
      levels.push(levelData);

      logger.debug(`[AdHeatmap] Level ${levelIndex}: ${levelData.adCount} ads, density: ${(levelData.adDensity * 100).toFixed(1)}%`);

      // Call optional callback for progressive content extraction
      if (onScrollLevel && typeof onScrollLevel === 'function') {
        try {
          await onScrollLevel(scrollY, viewportHeight, levelIndex);
        } catch (e) {
          logger.debug(`[AdHeatmap] onScrollLevel callback error at level ${levelIndex}:`, e.message);
        }
      }

      return levelData;
    });

    return this.analyzeHeatmap(levels);
  }

  analyzeHeatmap(levels) {
    if (levels.length === 0) {
      return {
        levels: [],
        totalScrollLevels: 0,
        totalAdsDetected: 0,
        avgAdDensity: 0,
        avgCLS: 0,
        adsAboveFold: 0,
        infiniteAdsPattern: false,
        scrollTrapDetected: false,
        mfaScore: 0
      };
    }

    const totalAds = levels.reduce((sum, l) => sum + l.adCount, 0);
    const avgDensity = levels.reduce((sum, l) => sum + l.adDensity, 0) / levels.length;
    const avgCLS = levels.reduce((sum, l) => sum + l.cls, 0) / levels.length;
    const adsAboveFold = levels[0]?.adsAboveFold || 0;

    // Detect infinite scroll MFA pattern (ads increase as you scroll)
    const adCounts = levels.map(l => l.adCount);
    const isInfiniteAdsPattern = levels.length > 3 &&
      adCounts.slice(-3).every((c, i, arr) => i === 0 || c >= arr[i - 1] * 0.8);

    // Detect scroll trap (very high ad density)
    const scrollTrapDetected = avgDensity > 0.25;

    // Ad distribution analysis
    const adDistribution = {
      top: levels.slice(0, Math.ceil(levels.length / 3))
        .reduce((sum, l) => sum + l.adCount, 0),
      middle: levels.slice(Math.ceil(levels.length / 3), Math.ceil(2 * levels.length / 3))
        .reduce((sum, l) => sum + l.adCount, 0),
      bottom: levels.slice(Math.ceil(2 * levels.length / 3))
        .reduce((sum, l) => sum + l.adCount, 0)
    };

    const analysis = {
      levels,
      totalScrollLevels: levels.length,
      totalAdsDetected: totalAds,
      avgAdDensity: avgDensity,  // Use the local variable avgDensity
      avgCLS,
      adsAboveFold,
      adDistribution,
      infiniteAdsPattern: isInfiniteAdsPattern,
      scrollTrapDetected,
      mfaScore: this.calculateMFAScore(levels, avgDensity, avgCLS, adsAboveFold)
    };

    logger.info('[AdHeatmap] Analysis complete', {
      totalAds,
      avgDensity: (avgDensity * 100).toFixed(1) + '%',
      avgCLS: avgCLS.toFixed(3),
      mfaScore: analysis.mfaScore
    });

    return analysis;
  }

  calculateMFAScore(levels, avgDensity, avgCLS, adsAboveFold) {
    let score = 0;

    // Ad density scoring (0-30 points)
    if (avgDensity > 0.4) score += 30;
    else if (avgDensity > 0.25) score += 20;
    else if (avgDensity > 0.15) score += 10;
    else if (avgDensity > 0.08) score += 5;

    // CLS scoring (0-25 points) - layout instability is MFA indicator
    if (avgCLS > 0.25) score += 25;
    else if (avgCLS > 0.1) score += 15;
    else if (avgCLS > 0.05) score += 8;

    // Ads above fold (0-20 points)
    if (adsAboveFold > 4) score += 20;
    else if (adsAboveFold > 2) score += 12;
    else if (adsAboveFold > 1) score += 5;

    // Infinite scroll pattern (0-25 points)
    const adCounts = levels.map(l => l.adCount);
    if (levels.length > 3) {
      const firstHalf = adCounts.slice(0, Math.floor(levels.length / 2));
      const secondHalf = adCounts.slice(Math.floor(levels.length / 2));
      const firstAvg = firstHalf.reduce((a, b) => a + b, 0) / firstHalf.length;
      const secondAvg = secondHalf.reduce((a, b) => a + b, 0) / secondHalf.length;

      if (secondAvg > firstAvg * 1.5) score += 25;
      else if (secondAvg > firstAvg * 1.2) score += 15;
    }

    return Math.min(100, score);
  }

  // Quick scan without full scroll
  async quickScan() {
    const ads = await this.detectAdsInViewport();
    const cls = await this.measureCLS();
    const dimensions = await this.getPageDimensions();

    return {
      adsVisible: ads.length,
      adsAboveFold: ads.filter(a => a.isAboveFold).length,
      cls,
      estimatedTotalLevels: Math.ceil(dimensions.totalHeight / dimensions.viewportHeight)
    };
  }
}

module.exports = AdHeatmapGenerator;

File: src/analyzers/ad_heatmap.py
"""
Ad Heatmap Generator - Per-scroll-level ad density analysis.
Industry-standard implementation with CLS measurement.
Ported from JS worker's ad-heatmap.js
"""

from typing import Any
import re

from src.utils.logger import get_logger

logger = get_logger(__name__)


# Comprehensive ad selectors (from JS worker - 90+ selectors)
AD_SELECTORS = [
    # Google Ads / AdX / Ad Manager / GPT
    '[id*="google_ads"]', '[id*="gpt-"]', '[class*="adunit"]',
    '[data-google-query-id]', 'ins.adsbygoogle', '[id*="div-gpt-ad"]',
    '[id*="google_ads_iframe"]', 'iframe[id*="google_ads"]',
    'iframe[src*="googleads"]', 'iframe[src*="tpc.googlesyndication"]',
    'iframe[src*="pagead2.googlesyndication"]', 'iframe[src*="securepubads"]',
    '[data-ad-slot]', '[data-ad-client]', '[data-ad-format]',
    '[class*="adsbygoogle"]', '[id*="aswift"]',
    
    # Google AdX / Ad Exchange specific
    'iframe[id*="aswift_"]', 'iframe[name*="aswift_"]',
    '[id*="google_image_div"]', 'div[id*="ad_unit"]',
    'iframe[src*="safeframe"]', '[class*="safeframe"]',
    
    # Generic ad containers
    '[class*="ad-slot"]', '[class*="advertisement"]', '[class*="ad-container"]',
    '[class*="ad-wrapper"]', '[class*="ad-banner"]', '[data-ad]',
    
    # Native ads
    '[class*="taboola"]', '[class*="outbrain"]', '[class*="mgid"]',
    '[class*="revcontent"]', '[id*="taboola"]', '[id*="outbrain"]',
    
    # Video ads
    '[class*="video-ad"]', '[class*="preroll"]', '[class*="midroll"]',
    
    # Generic patterns
    '[id^="ad-"]', '[id^="banner-"]', '[id^="div-gpt-"]',
    '[class^="ad-"]', '[class^="banner-"]', '[class*="sponsored"]',
    '[aria-label="Advertisement"]', '[aria-label="Sponsored"]',
    
    # Specific networks
    '[id*="criteo"]', '[class*="criteo"]',
    '[id*="pubmatic"]', '[class*="pubmatic"]',
    '[id*="amazon"]', '[class*="amzn"]',
    
    # Deceptive patterns (MFA favorite)
    '[class*="download"]', '[id*="download"]',
    '[class*="fake-download"]', '[class*="dl-button"]',
    
    # Pop-under / overlay
    '[class*="overlay-ad"]', '[class*="popup-ad"]', '[class*="modal-ad"]',
    '[class*="interstitial"]', '[class*="splash-ad"]',
    
    # AdSense alternatives (MFA common)
    '[class*="propeller"]', '[id*="propeller"]',
    '[class*="adsterra"]', '[id*="adsterra"]',
    '[class*="monetag"]', '[id*="monetag"]',
]

# Deceptive text patterns
DECEPTIVE_PATTERNS = [
    re.compile(r"download\s*(now|free|here|button)", re.IGNORECASE),
    re.compile(r"click\s*(here|to|now)", re.IGNORECASE),
    re.compile(r"free\s*(download|install|get)", re.IGNORECASE),
    re.compile(r"install\s*(now|free)", re.IGNORECASE),
    re.compile(r"update\s*(required|now|available)", re.IGNORECASE),
    re.compile(r"your\s*(system|computer|device)", re.IGNORECASE),
    re.compile(r"virus\s*(detected|found|alert)", re.IGNORECASE),
    re.compile(r"warning[:\s]", re.IGNORECASE),
    re.compile(r"congratulations", re.IGNORECASE),
]


class AdHeatmapGenerator:
    """
    Generates per-scroll-level ad density heatmaps.
    
    Features:
    - Per-viewport ad density calculation
    - Ad distribution analysis (top/middle/bottom)
    - CLS (Cumulative Layout Shift) tracking
    - Infinite scroll pattern detection
    - Scroll trap detection
    - Deceptive ad pattern detection
    """
    
    def __init__(self):
        self.levels: list[dict[str, Any]] = []
    
    def analyze(
        self,
        ad_elements: list[dict[str, Any]],
        page_dimensions: dict[str, int],
        viewport_height: int = 1080,
    ) -> dict[str, Any]:
        """
        Analyze ad distribution across the page.
        
        Args:
            ad_elements: List of ad elements with positions
            page_dimensions: {totalHeight, viewportHeight, viewportWidth}
            viewport_height: Height of one viewport
            
        Returns:
            Heatmap analysis with MFA score
        """
        if not ad_elements:
            return self._empty_result()
        
        total_height = page_dimensions.get("totalHeight", viewport_height)
        viewport_width = page_dimensions.get("viewportWidth", 1920)
        
        # Create scroll levels
        num_levels = max(1, int(total_height / viewport_height))
        levels = []
        
        for i in range(min(num_levels, 10)):  # Max 10 levels
            level_top = i * viewport_height
            level_bottom = level_top + viewport_height
            
            # Find ads in this level
            level_ads = [
                ad for ad in ad_elements
                if self._ad_in_range(ad, level_top, level_bottom)
            ]
            
            # Calculate density
            viewport_area = viewport_height * viewport_width
            total_ad_area = sum(
                self._get_ad_area(ad)
                for ad in level_ads
            )
            ad_density = total_ad_area / viewport_area
            
            levels.append({
                "level_index": i,
                "scroll_y": level_top,
                "viewport_height": viewport_height,
                "ad_count": len(level_ads),
                "total_ad_area": total_ad_area,
                "ad_density": round(ad_density, 4),
                "ads_above_fold": len([a for a in level_ads if a.get("isAboveFold")]) if i == 0 else 0,
            })
        
        self.levels = levels
        return self._analyze_heatmap(levels, ad_elements)
    
    def _ad_in_range(
        self,
        ad: dict[str, Any],
        top: int,
        bottom: int,
    ) -> bool:
        """Check if ad is within scroll range."""
        ad_top = ad.get("y", 0)
        ad_bottom = ad_top + ad.get("height", 0)
        return ad_bottom > top and ad_top < bottom
    
    def _get_ad_area(self, ad: dict[str, Any]) -> int:
        """Get ad area in pixels."""
        return ad.get("width", 0) * ad.get("height", 0)
    
    def _analyze_heatmap(
        self,
        levels: list[dict[str, Any]],
        all_ads: list[dict[str, Any]],
    ) -> dict[str, Any]:
        """Analyze heatmap patterns."""
        if not levels:
            return self._empty_result()
        
        total_ads = sum(l["ad_count"] for l in levels)
        avg_density = sum(l["ad_density"] for l in levels) / len(levels)
        ads_above_fold = levels[0].get("ads_above_fold", 0) if levels else 0
        
        # Detect infinite scroll MFA pattern
        ad_counts = [l["ad_count"] for l in levels]
        infinite_ads_pattern = (
            len(levels) > 3 and
            all(ad_counts[i] >= ad_counts[i-1] * 0.8 for i in range(max(1, len(ad_counts)-3), len(ad_counts)))
        )
        
        # Detect scroll trap (very high ad density)
        scroll_trap_detected = avg_density > 0.25
        
        # Ad distribution analysis
        third = max(1, len(levels) // 3)
        ad_distribution = {
            "top": sum(l["ad_count"] for l in levels[:third]),
            "middle": sum(l["ad_count"] for l in levels[third:2*third]),
            "bottom": sum(l["ad_count"] for l in levels[2*third:]),
        }
        
        # Detect deceptive ads
        deceptive_ads = self._detect_deceptive_ads(all_ads)
        
        # Calculate MFA score
        mfa_score = self._calculate_mfa_score(
            levels=levels,
            avg_density=avg_density,
            ads_above_fold=ads_above_fold,
            deceptive_count=len(deceptive_ads),
        )
        
        return {
            "levels": levels,
            "total_scroll_levels": len(levels),
            "total_ads_detected": total_ads,
            "avg_ad_density": round(avg_density, 4),
            "ads_above_fold": ads_above_fold,
            "ad_distribution": ad_distribution,
            "infinite_ads_pattern": infinite_ads_pattern,
            "scroll_trap_detected": scroll_trap_detected,
            "deceptive_ads": deceptive_ads,
            "deceptive_ad_count": len(deceptive_ads),
            "mfa_score": mfa_score,
            "risk_level": self._get_risk_level(mfa_score),
        }
    
    def _detect_deceptive_ads(
        self,
        ads: list[dict[str, Any]],
    ) -> list[dict[str, Any]]:
        """Detect deceptive ad patterns."""
        deceptive = []
        
        for ad in ads:
            text = ad.get("text", "") or ad.get("innerText", "")
            selector = ad.get("selector", "")
            
            # Check for deceptive text
            for pattern in DECEPTIVE_PATTERNS:
                if pattern.search(text):
                    deceptive.append({
                        "type": "deceptive_text",
                        "text": text[:50],
                        "pattern": pattern.pattern,
                    })
                    break
            
            # Check for download button patterns
            if "download" in selector.lower() or "download" in text.lower():
                if any(word in text.lower() for word in ["free", "now", "click"]):
                    deceptive.append({
                        "type": "fake_download",
                        "selector": selector[:50],
                    })
        
        return deceptive
    
    def _calculate_mfa_score(
        self,
        levels: list[dict[str, Any]],
        avg_density: float,
        ads_above_fold: int,
        deceptive_count: int,
    ) -> int:
        """Calculate MFA score from heatmap (0-100)."""
        score = 0
        
        # Ad density scoring (0-30 points)
        if avg_density > 0.4:
            score += 30
        elif avg_density > 0.25:
            score += 20
        elif avg_density > 0.15:
            score += 10
        elif avg_density > 0.08:
            score += 5
        
        # Ads above fold (0-20 points)
        if ads_above_fold > 4:
            score += 20
        elif ads_above_fold > 2:
            score += 12
        elif ads_above_fold > 1:
            score += 5
        
        # Infinite scroll pattern (0-25 points)
        if len(levels) > 3:
            ad_counts = [l["ad_count"] for l in levels]
            first_half = ad_counts[:len(ad_counts)//2]
            second_half = ad_counts[len(ad_counts)//2:]
            
            if first_half and second_half:
                first_avg = sum(first_half) / len(first_half)
                second_avg = sum(second_half) / len(second_half)
                
                if second_avg > first_avg * 1.5:
                    score += 25
                elif second_avg > first_avg * 1.2:
                    score += 15
        
        # Deceptive ads (0-25 points)
        if deceptive_count > 3:
            score += 25
        elif deceptive_count > 1:
            score += 15
        elif deceptive_count > 0:
            score += 8
        
        return min(100, score)
    
    def _get_risk_level(self, score: int) -> str:
        """Map score to risk level."""
        if score >= 60:
            return "high"
        elif score >= 30:
            return "medium"
        else:
            return "low"
    
    def _empty_result(self) -> dict[str, Any]:
        """Return empty heatmap result."""
        return {
            "levels": [],
            "total_scroll_levels": 0,
            "total_ads_detected": 0,
            "avg_ad_density": 0,
            "ads_above_fold": 0,
            "ad_distribution": {"top": 0, "middle": 0, "bottom": 0},
            "infinite_ads_pattern": False,
            "scroll_trap_detected": False,
            "deceptive_ads": [],
            "deceptive_ad_count": 0,
            "mfa_score": 0,
            "risk_level": "low",
        }


# Convenience instance
ad_heatmap_generator = AdHeatmapGenerator()
